docs: add a contributing Git cheatsheet document

anandkaranubc opened the pull request:
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request:

- Adds a Git cheatsheet to the contributing documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

- resolves none

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
docs: add a Git cheatsheet for stdlib

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
chore: fix linting issues

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
Here’s more documentation! It’s hard to classify this as a 'cheat sheet'—it could probably be expanded into a more comprehensive document.

I haven’t included anything about resolving merge conflicts yet. Any thoughts on that?

anandkaranubc commented:
Is it also a good idea to add these external links (at the end)?

https://education.github.com/git-cheat-sheet-education.pdf
https://training.github.com/downloads/github-git-cheat-sheet.pdf

kgryte commented on docs/contributing/git_cheatsheet.md:
````diff
@@ -0,0 +1,308 @@
+<!-- run-disable -->
+
+```bash
+git commit -m "feat: add support for new function"
````
@anandkaranubc I think it worthwhile to also include a commit example using `git commit`. Multi-line commits have tripped people up in the past.

kgryte commented on docs/contributing/git_cheatsheet.md:
```diff
@@ -0,0 +1,308 @@
+
+# Git Cheatsheet
+
+> Git Cheatsheet for stdlib
```
```suggestion
> A cheatsheet for using Git to develop stdlib.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (commented):
This looks great, @anandkaranubc! Let a comment about also giving folks a hint/demo of a multi-line commit which opens their terminal editor.

kgryte commented:
> I haven’t included anything about resolving merge conflicts yet. Any thoughts on that?

Yes, it may be good to demonstrate what a conflict looks like (e.g., `<<<<<`, etc). And then how to rectify.

> Is it also a good idea to add these external links (at the end)?

Yes, you can include a "Other References" section at the end which contains a list of links.

> Other things?

- Could also mention the use of `make commit` which provides an interactive prompt to ensure adherence to project Git conventions.
- May be worthwhile demoing an actual workflow as a series of Git commands (e.g., checkout develop, pull latest changes from upstream, push to origin, create new feature branch, add commits, rebase on top of latest develop, push new branch to origin, pull latest changes from origin (e.g., after having a reviewer commit directly to your branch), add additional commits, merge in latest develop due to having already pushed changes to remote, push changes to origin, etc).

anandkaranubc committed:
docs: add a section for conflicts and example workflow

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
docs: add other reference links

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
@kgryte I hope you enjoy the read!

kgryte commented on docs/contributing/git_cheatsheet.md:
````diff
@@ -0,0 +1,515 @@
+git reset --hard upstream/develop
+```
+
+> **Warning:** This will delete any changes you made to `develop`. Make sure you don’t have important work in this branch before running this command.
````
```suggestion
> **Warning:** This will delete any changes you made to `develop`. Make sure you don't have important work in this branch before running this command.
```

kgryte committed:
style: remove fancy quote

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on docs/contributing/git_cheatsheet.md:
```diff
@@ -0,0 +1,515 @@
+
+> **Tip:** Many editors like VS Code highlight conflicts and even give you buttons to accept incoming or current changes. This can make resolving conflicts much easier.
+
+Merge conflicts can be annoying, but they’re a natural part of working with others. Take your time, and don’t hesitate to ask for help if you're stuck.
```
```suggestion
Merge conflicts can be annoying, but they're a natural part of working with others. Take your time, and don't hesitate to ask for help if you're stuck.
```

kgryte committed:
style: remove fancy quotes

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM. Any further changes can be addressed in follow-up PRs.
